MINOR: Document Offset and Partition 0-indexing, fix typo

Scanteianu opened the pull request:
### Summary

Add comments to clarify that both offsets and partitions are 0-indexed, and fix a minor typo. Clarify which offset will be retrieved by `poll()` after `seek()` is used in various circumstances.

Add tests for mock consumer to illustrate behaviour described in new seek documentation. It is assumed that the behaviour of `MockConsumer` is in sync with `KafkaConsumer` with respect to `seek()` behaviour (if it ever becomes out of sync, it will be a bug, as a good mock should accurately model the behaviour of the real component).

### Tests
Test added for mock consumer.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Scanteianu committed:
add comments about 0-indexing of the partitions and the offsets

Scanteianu committed:
fix unrelated typo

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,6 +1581,10 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
      * is invoked for the same partition more than once, the latest offset will be used on the next poll(). Note that
      * you may lose data if this API is arbitrarily used in the middle of consumption, to reset the fetch offsets
      *
+     * Note that, since messages are 0-indexed, {@link #seekToBeginning(Collection)} should be used to retrieve
+     * the message at offset 0.
+     *
+     * @param offset the last offset which was consumed. The next message received will have a higher offset number than this
```
I'm not sure this is correct. We have many tests relying on seek(tp, 0) to make sure it starts fetching from offset 0. Ex: `BaseConsumerTest#testSimpleConsumption`. Could you find or write a simple test to confirm what you described here? Maybe there's a bug. Thanks.

Scanteianu replied:
Thank you very much, you were correct. I have added a few tests in mock consumer meant to confirm the behaviour. These work on the assumption that the behaviour for the real consumer is the same as for the mock consumer (the mock consumer was just easier to write tests for)

showuon reviewed (commented):
Thanks for the patch. Left a comment.

Scanteianu committed:
add tests for mock consumer to clarify seek behaviour, fix comment

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

Scanteianu committed:
clarify documentation wording for seek

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,6 +1581,19 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * specified but there is one with a higher offset.
+     * seek(0) is equivalent to seek to beginning for a topic with beginning offset 0.
+     *
+     * Seeking past the end of the highest known offset means that all consumer records between the highest known offset
+     * and the offset passed to seek will be skipped by poll(), which will only return records once the offset passed
+     * to poll() has been reached.
```
Are you sure this is correct? Won't it return `OffsetOutOfRangeException` and reset the offset by `auto.offset.reset` config?

showuon reviewed (commented):
@Scanteianu , thanks for the update, but there is still something wrong. For the test, I don't think another tests in MockConsumer is needed and correct. After all, MockConsumer is a `Mock` one, some behaviors doesn't match the real consumer. For the seek test, we already have one here: `PlaintextConsumerTest#testSeek`. If you think there are some tests missing, you can add in the test. 

BTW, `PlaintextConsumerTest` is integration tests, so it'll launch real brokers and consumers to do tests. You can test it there. 

Thanks.

Scanteianu commented:
> @Scanteianu , thanks for the update, but there is still something wrong. For the test, I don't think another tests in MockConsumer is needed and correct. After all, MockConsumer is a `Mock` one, some behaviors doesn't match the real consumer. For the seek test, we already have one here: `PlaintextConsumerTest#testSeek`. If you think there are some tests missing, you can add in the test.
> 
> BTW, `PlaintextConsumerTest` is integration tests, so it'll launch real brokers and consumers to do tests. You can test it there.
> 
> Thanks.

@showuon thanks for the review, I really appreciate your taking the time to look at it, and point me to the right places in the code to add tests and the configs that specify behaviour!

I have worked with people who believe mocks should be as close to the behaviour of the "real deal" as possible (and am partial to this opinion myself), but I'll maybe open another PR with that, as I think that is more of a stylistic opinion (whether Apache has this codified somewhere, I don't know), and I'd really like to improve the seek documentation, as I spent quite a few hours googling for this stuff. 

For now, I've moved the tests to the Plaintext Consumer, and added tests for the various auto.offset.reset configs. The behaviours in latest seem super weird to me, based on your explanation - they don't seem to match https://medium.com/lydtech-consulting/kafka-consumer-auto-offset-reset-d3962bad2665 specifically for the `latest` value. 

In the case of `latest`, if the next batch of records contains the offset that is passed to seek, it seems to skip to that offset, which is the same behaviour that the mock consumer had, which kind of makes sense. However, if the next batch of records doesn't contain the offset last passed to seek, it resets to the very beginning (offset 0). Either I am missing something, or this might be a bug?

Scanteianu committed:
remove mock consumer test, add test to plaintext consumer

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

Scanteianu commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,73 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+    // assert that poll resets to the ending position
+    assertTrue(consumer.poll(Duration.ofMillis(50)).isEmpty)
+    sendRecords(producer, totalRecords.toInt, tp, startingTimestamp = totalRecords)
+    // ensure that new records are consumed only if they're after the seek position
```
not sure if this is desired, or simply current behaviour - please review

Scanteianu commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,73 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+    assertTrue(consumer.poll(Duration.ofMillis(50)).isEmpty)
+    sendRecords(producer, totalRecords.toInt, tp, startingTimestamp = totalRecords)
+    // why is this resetting to 0? this seems like a bug
+    consumeAndVerifyRecords(consumer = consumer, numRecords = 1, startingOffset = 0, startingTimestamp = 0)
```
this definitely seems wrong (I left the test in to demonstrate that it happens - the test passes), please review

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,73 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+
+  @Test
+  def testFetchInvalidOffsetResetConfigLatest(): Unit = {
+    val consumer = createConsumer()
+    this.consumerConfig.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest")
```
The consumer config should be fed into `createConsumer` method. You can refer to here for how to set config for test consumer: https://github.com/apache/kafka/blob/trunk/core/src/test/scala/integration/kafka/api/AbstractConsumerTest.scala#L100-L102

Scanteianu replied:
thanks! I was following the example in `testFetchInvalidOffset` but you're right I got the order wrong here. I don't know how i missed that I updated the config after building the consumer

Scanteianu replied:
so i fixed this, now it seems like the behaviour is much more predictable, but it still feels like the first `poll` after `seek` to invalid offset should reset to the current latest unconsumed offset and/or the end of the queue (ieL offset 10). Nonetheless, this does seem like it might be a conscious design decision, and therefore all the more important to document.

Scanteianu committed:
fix seek to invalid with offset reset config latest tests, and clarify documentation

Scanteianu commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,76 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+    val outOfRangePos = totalRecords + 17 //arbitrary, much higher offset
+    consumer.seek(tp, outOfRangePos)
+    // assert that poll resets to the ending position
+    assertTrue(consumer.poll(Duration.ofMillis(50)).isEmpty)
```
shouldn't this reset the offset to 10, so the next poll after new records should return offset 10?

Scanteianu replied:
i guess it might make sense that upon the arrival of new records, consumer tries to seek, and if that fails only then does it reset to the latest record in the new set of records

Scanteianu commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,76 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+    val outOfRangePos = totalRecords + 7 //arbitrary, much higher offset
+    consumer.seek(tp, outOfRangePos)
+    // assert that poll resets to the ending position
+    assertTrue(consumer.poll(Duration.ofMillis(50)).isEmpty)
```
shouldn't this reset the offset to 10, so the next poll after new records should return offset 10?

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

Scanteianu commented:
@showuon could I get another review please

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,76 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+  def testFetchInvalidOffsetResetConfigLatest(): Unit = {
+
+    this.consumerConfig.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest")
+    val consumer = createConsumer(configOverrides = this.consumerConfig)
```
Try adding one more custom config, you'll get what you expected:
`this.consumerConfig.setProperty(ConsumerConfig.FETCH_MAX_WAIT_MS_CONFIG, "0")`

showuon replied:
Some explanation:
To reduce consumer/follower fetch overhead, we will wait for a while in broker side if there's no records for a specific offset, to reduce the network traffic from client/broker. Ex: in your example, you send 10 records to broker, after received 10 records (offset 0 -9), the next consumer poll will try to fetch offset 10, but there's no offset 10 record in broker. So, when broker receiving the FETCH request, it'll wait in broker side for `fetch.max.wait.ms` (default 500 ms). If there's records arrived within 500ms, it'll respond to client. You can imagine, that reduce the network traffic a lot. And, if after 500ms, there's no record arrived, it'll respond with empty result. So, your current test will run like this:

1. sending 10 records to broker
2. consumer poll 10 records + one more fetch for offset 10
3. consumer seek to 17
4. consumer poll, but not sending FETCH request because there's already an in-flight FETCH request not responded, yet
5. send another 10 records
6. consumer poll, the previous FETCH responded with 10 records, starting from offset 10. But they got discarded because fetcher found the starting offset is not what we expected, so send another FETCH staring from offset 17
7. received records starting from offset 17

If we set the  `fetch.max.wait.ms` to 0, it'll be like this:
1. sending 10 records to broker
2. consumer poll 10 records + one more fetch for offset 10, **and responded with empty results immediately**
3. consumer seek to 17
//4. consumer poll, but not sending FETCH request because there's already an in-flight FETCH request not responded, yet
4. consumer poll, sending FETCH request, got out of range error, reset position to latest, 10
5. send another 10 records
//6. consumer poll, it'll get 10 records, but got discarded because fetcher found the starting offset is not what we expected, so send another FETCH staring from offset 17
6. consumer poll, got 10 records
7. received records starting from offset **10**

Hope that's clear

Scanteianu replied:
Thank you so much, I will update both the tests and the documentation.

Scanteianu replied:
this is an AMAZING explanation, and I really appreciate you taking the time to help me understand what's going on. 

My hope with the proposal in KIP-886 is that this should be easier for someone to discover as they are building a producer or consumer, especially if the documentation is good for setting configs like `AUTO_OFFSET_RESET_CONFIG` (ie: it can mention that "latest" interacts with the `FETCH_MAX_MS_WAIT_CONFIG`)

Scanteianu committed:
add FETCH_MAX_WAIT_MS_CONFIG to tests, and document its behaviour

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,77 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+  }
+
+  @Test
+  def testFetchInvalidOffsetResetConfigLatestRecordsBeforeTimeout(): Unit = {
```
I don't think this test is testing reset offset logic. Could we remove this test?

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,77 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+  }
+
+  @Test
+  def testFetchInvalidOffsetResetConfigLatestNoRecordsBeforeTimeout(): Unit = {
```
If you agree to remove the above test, this test could be simply renamed to `testFetchOutOfRangeOffsetResetConfigLatest()`. WDYT?

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,77 @@ class PlaintextConsumerTest extends BaseConsumerTest {
   }
 
+  @Test
+  def testFetchInvalidOffsetResetConfigEarliest(): Unit = {
```
nit: Since it's not an invalid offset, it's out of range offset, could we rename to `testFetchOutOfRangeOffsetResetConfigEarliest`?

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,77 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+  @Test
+  def testFetchInvalidOffsetResetConfigEarliest(): Unit = {
+    this.consumerConfig.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
+    val consumer = createConsumer(configOverrides = this.consumerConfig)
```
Although everything works well now, I think we should still add `this.consumerConfig.setProperty(ConsumerConfig.FETCH_MAX_WAIT_MS_CONFIG, "0")` for earliest test, to make sure the offset reset is working as expected

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,77 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+  @Test
+  def testFetchInvalidOffsetResetConfigLatestNoRecordsBeforeTimeout(): Unit = {
+    this.consumerConfig.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest")
+    this.consumerConfig.setProperty(ConsumerConfig.FETCH_MAX_WAIT_MS_CONFIG, "0")
```
Could we add a comment above this line to mention why we need to set `FETCH_MAX_WAIT_MS_CONFIG` to 0?

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,77 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+    sendRecords(producer, totalRecords.toInt, tp, startingTimestamp = startingTimestamp)
+    consumer.assign(List(tp).asJava)
+    consumer.seek(tp, 0)
+    //consume some, but not all of the records
```
Please add a space after `//`

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,77 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+    assertTrue(consumer.poll(Duration.ofMillis(50)).isEmpty)
+    sendRecords(producer, totalRecords.toInt, tp, startingTimestamp = totalRecords)
+    val nextRecord = consumer.poll(Duration.ofMillis(50)).iterator().next()
+    //ensure the seek went to the last known record at the time of the previous poll
```
Please add a space after `//`

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,77 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+    sendRecords(producer, totalRecords.toInt, tp, startingTimestamp = totalRecords)
+    val nextRecord = consumer.poll(Duration.ofMillis(50)).iterator().next()
+    //ensure the seek went to the last known record at the time of the previous poll
+    assertEquals(totalRecords,nextRecord.offset())
```
Please add a space after comma, i.e. `assertEquals(totalRecords, nextRecord.offset())`

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,77 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+    //consume some, but not all of the records
+    consumeAndVerifyRecords(consumer = consumer, numRecords = totalRecords.toInt/2, startingOffset = 0)
+    // seek to out of range position
+    val outOfRangePos = totalRecords + 17 //arbitrary, much higher offset
```
Please add a space after `//`

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,6 +1581,23 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * by default, with {@link ConsumerConfig AUTO_RESET_CONFIG} set to "latest"
+     * a higher numbered offset. A higher numbered offset will be returned if there is no consumer record with the offset
+     * specified but there is one with a higher offset.
+     * seek(0) is equivalent to seek to beginning for a topic with beginning offset 0.
```
This is not always correct. The start offset might not be 0 if it got deleted manually or exceeding retention time/size.... etc
* {@link #seekToBeginning(Collection)} is equivalent to seek to beginning ...

Scanteianu replied:
added a clarification, I think it's worth having the note because it clarifies that records are 0 indexed at the creation of a topic

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -672,7 +672,7 @@ class PlaintextConsumerTest extends BaseConsumerTest {
   def testFetchInvalidOffset(): Unit = {
     this.consumerConfig.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "none")
-    val consumer = createConsumer()
+    val consumer = createConsumer(configOverrides = this.consumerConfig)
```
Nice fix!

Scanteianu replied:
i think that `createConsumer()` basically does this anyways (it reads `this.consumerConfig` and passes it through, but i agree this is more explicit)

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,6 +1581,23 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
      * is invoked for the same partition more than once, the latest offset will be used on the next poll(). Note that
      * you may lose data if this API is arbitrarily used in the middle of consumption, to reset the fetch offsets
      *
+     * The next Consumer Record which will be retrieved when poll() is invoked will either have the offset specified or,
+     * by default, with {@link ConsumerConfig AUTO_RESET_CONFIG} set to "latest"
+     * a higher numbered offset. A higher numbered offset will be returned if there is no consumer record with the offset
+     * specified but there is one with a higher offset.
```
It's hard to read for these 4 lines. I don't know why we should mention "latest" reset config here.

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,6 +1581,23 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * seek(0) is equivalent to seek to beginning for a topic with beginning offset 0.
+     * seekToEnd() is equivalent to seeking to the highest known offset + 1.
+     *
+     * Seeking past the end of the highest known offset means an invalid offset is reached.
```
Seeking to the offset earlier than start offset will also return out of range exception. So, maybe update to:
`Seeking to the offset smaller than the log start offset or larger than the log end offset or high watermark means an invalid offset is reached`

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,6 +1581,23 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     *
+     * Seeking past the end of the highest known offset means an invalid offset is reached.
+     * Invalid offset behaviour is controlled by
+     * the {@link ConsumerConfig AUTO_RESET_CONFIG} property. If this is set to "earliest", the next poll() will seek to the beginning
+     * before returning a record. If it is set to "latest", it will seek to the last known record (similar to seekToEnd()).
```
seek will happen right after you call `seek`, it doesn't happen after calling `poll`. How about this:
`If this is set to "earliest", the next poll will return records from the starting offset. `

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,6 +1581,23 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * Invalid offset behaviour is controlled by
+     * the {@link ConsumerConfig AUTO_RESET_CONFIG} property. If this is set to "earliest", the next poll() will seek to the beginning
+     * before returning a record. If it is set to "latest", it will seek to the last known record (similar to seekToEnd()).
+     * Note that, in the "latest" if any new records show up before {@link ConsumerConfig FETCH_MAX_WAIT_MS_CONFIG}, then
+     * the next poll() may succeed in seeking to the requested offset, or otherwise it will seek to the new latest known offset.
```
I'm thinking we don't need to mention it here. First, it's not just `latest` case, `earlier` will also has this behavior. Second, it's too detail to users to know about it, and make users more confused about it, like so, how should we do about it? Third, so, back to the point you tried to mentioned, I think you're trying to explain the case:
1. current log end offset is 10
2. seek to 17
3. poll -> got nothing
4. produce 10 more records
5. poll -> got records offset starting from 17

Is that right? I think that's expected from user's point of view. User tried to seek to offset 17 and returning records starting from 17, everything works fine, isn't it? 

I think if you really want to talk about it, maybe you can say: the seek offset won't change to the in-flight fetch request, it will take effect in next fetch request. 

WDYT?

IMO, we don't need to mention it.

Scanteianu replied:
i think this makes sense, but I'd like to leave in the note about fetch_max_wait_ms because that confused me when I was writing the tests

showuon reviewed (commented):
@Scanteianu , thanks for the update. Left some comments.

Scanteianu committed:
address comments - remove a latest test, rename another, and refactor documentation for seek

Scanteianu committed:
add spaces before comments

Scanteianu committed:
add even more spaces

Scanteianu committed:
clarify seek(0) note

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,6 +1581,26 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * a record with that offset exists (ie: it is a valid offset).
+     *
+     * seek(0) is equivalent to seek to beginning for a topic with beginning offset 0,
+     * assuming that there is a record at offset 0 still available.
```
I'm fine to keep `seek(0)` here, but I think we should put `seekToBeginning`, too. It should be the correct method if you want to seek back to the beginning, no matter if the start offset is 0 or not. Maybe:
`seekToEnd() is equivalent to seek to beginning for a topic partition. seek(0) also works if the current start offset is 0.`
WDYT?

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,6 +1581,26 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * If it is set to "latest", it will seek to the last known record (similar to seekToEnd()).
+     * If it is set to "none", an {@link OffsetOutOfRangeException} will be thrown.
+     *
+     * Note that,  the seek offset won't change to the in-flight fetch request, it will take effect in next fetch request.
```
additional space: Note that, [] the seek offset ....

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,6 +1581,26 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     *
+     * Note that,  the seek offset won't change to the in-flight fetch request, it will take effect in next fetch request.
+     * In order for the invalid offset behaviour to take effect immediately after seeking and polling,
+     * {@link ConsumerConfig FETCH_MAX_WAIT_MS_CONFIG} should be set to 0.
```
1. Please change `{@link ConsumerConfig FETCH_MAX_WAIT_MS_CONFIG}` to `{@code fetch.max.wait.ms}`
2. I don't think we should recommend users set fetch.max.wait.ms=0 here. We just need to explain why it won't take effect immediately. 

Maybe:
`Note that, the seek offset won't change to the in-flight fetch request, it will take effect in next fetch request. So, it might wait for {@code fetch.max.wait.ms} before starting fetching the records from desired offset.`

WDYT?

showuon reviewed (commented):
Thanks for the update. Some comments left.

Scanteianu committed:
further rewording, extra space

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

Scanteianu committed:
add <p> tags for better javadoc format

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,7 +1581,27 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * The next Consumer Record which will be retrieved when poll() is invoked will have the offset specified, given that
+     * a record with that offset exists (ie: it is a valid offset).
+     * <p>
+     * {@link KafkaConsumer#seekToBeginning(Collection)} will go to the first offset in the topic.
```
nit: `{@link KafkaConsumer#seekToBeginning(Collection)}` -> {@link #seekToBeginning(Collection)}

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,7 +1581,27 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
      * you may lose data if this API is arbitrarily used in the middle of consumption, to reset the fetch offsets
+     * <p>
+     * The next Consumer Record which will be retrieved when poll() is invoked will have the offset specified, given that
+     * a record with that offset exists (ie: it is a valid offset).
```
nit: `ie.` -> `i.e.`

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,7 +1581,27 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * a record with that offset exists (ie: it is a valid offset).
+     * <p>
+     * {@link KafkaConsumer#seekToBeginning(Collection)} will go to the first offset in the topic.
+     * seek(0) is equivalent to seekToBeginning for a topic with beginning offset 0,
```
for a topic [partition] with beginning offset 0

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,7 +1581,27 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * {@link KafkaConsumer#seekToBeginning(Collection)} will go to the first offset in the topic.
+     * seek(0) is equivalent to seekToBeginning for a topic with beginning offset 0,
+     * assuming that there is a record at offset 0 still available.
+     * {@link KafkaConsumer#seekToEnd(Collection)} is equivalent to seeking to the highest known offset + 1.
```
nit: `{@link #seekToEnd(Collection)}`

showuon replied:
is equivalent to `seek` to the highest known offset + 1

showuon replied:
Could we say:
{seekToEnd(Collection)} is equivalent to seeking to `the highest known offset + 1` -> {seekToEnd(Collection)} is equivalent to seeking to `the last offset of the partition`. 
That's the wording in the javadoc in `seekToEnd` method.

showuon replied:
One thing to note here, the seekToEnd might seek to `log end offset` or `Last Stable Offset` depends on `read-committed` or `read-uncommitted`. So we need to be cautious for the wording here.

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,7 +1581,27 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * Seeking to the offset smaller than the log start offset or larger than the log end offset
+     * or high watermark means an invalid offset is reached.
+     * Invalid offset behaviour is controlled by
+     * the {@link ConsumerConfig AUTO_RESET_CONFIG} property.
```
nit: These 2 lines can be put in the same line
`Invalid offset behaviour is controlled by`
`the {@link ConsumerConfig AUTO_RESET_CONFIG} property.`

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,7 +1581,27 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * Invalid offset behaviour is controlled by
+     * the {@link ConsumerConfig AUTO_RESET_CONFIG} property.
+     * If this is set to "earliest", the next poll will return records from the starting offset.
+     * If it is set to "latest", it will seek to the last known record (similar to seekToEnd()).
```
nit: it will seek to the last `known record` -> it will seek to the last `offset`

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,7 +1581,27 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * the {@link ConsumerConfig AUTO_RESET_CONFIG} property.
+     * If this is set to "earliest", the next poll will return records from the starting offset.
+     * If it is set to "latest", it will seek to the last known record (similar to seekToEnd()).
+     * If it is set to "none", an {@link OffsetOutOfRangeException} will be thrown.
```
I don't think it will link to `OffsetOutOfRangeException` with this format. Maybe we just use `code`:
* If it is set to "none", an {@code OffsetOutOfRangeException} will be thrown.

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,7 +1581,27 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * Note that, the seek offset won't change to the in-flight fetch request, it will take effect in next fetch request.
+     * So, the consumer might wait for {@code fetch.max.wait.ms} before starting to fetch the records from desired offset.
      *
+     * @param offset the next offset returned by poll() will be either this or greater.
```
Since we're here, should we also add param for `partition` ?

showuon reviewed (commented):
@Scanteianu , thanks for the update. Left some comments.

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

Scanteianu committed:
address nits from @showuon

Scanteianu commented:
> @Scanteianu , thanks for the update. Left some comments.

thanks for the comments @showuon , addressed!

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,7 +1581,28 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * seek(0) is equivalent to seekToBeginning for a TopicPartition with beginning offset 0,
+     * assuming that there is a record at offset 0 still available.
+     * {@link #seekToEnd(Collection)} is equivalent to seeking to the last offset of the partition, but behavior depends on
+     * {@code isolation.level}, so see {@link #seekToBeginning(Collection)} documentation for more details.
```
I think you want to refer to `{@link #seekToEnd(Collection)}`, not `{@link #seekToBeginning(Collection)}`, right?

Scanteianu replied:
thanks for this, good catch

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,7 +1581,28 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * {@link #seekToEnd(Collection)} is equivalent to seeking to the last offset of the partition, but behavior depends on
+     * {@code isolation.level}, so see {@link #seekToBeginning(Collection)} documentation for more details.
+     * <p>
+     * Seeking to the offset smaller than the log start offset or larger than the log end offset
+     * or high watermark means an invalid offset is reached.
```
The term needs to be updated. How about:
     * Seeking to the offset smaller than the log start offset or larger than the last offset means an invalid offset is reached.

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,7 +1581,28 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * <p>
+     * Seeking to the offset smaller than the log start offset or larger than the log end offset
+     * or high watermark means an invalid offset is reached.
+     * Invalid offset behaviour is controlled by the {@link ConsumerConfig AUTO_RESET_CONFIG} property.
```
`{@link ConsumerConfig AUTO_RESET_CONFIG}` should be updated to `{@code auto.offset.reset}`, because the link won't work with this format.

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1581,7 +1581,28 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+     * So, the consumer might wait for {@code fetch.max.wait.ms} before starting to fetch the records from desired offset.
      *
+     * @param partition The TopicPartition on which the seek will be performed.
+     * @param offset the next offset returned by poll() will be either this or greater.
```
About the `greater`, I don't know what you are referring to. Could you elaborate more? 
I think it's possible we returned greater offset if this is a compact topic. But I don't think we need to make this line that complicated.

Scanteianu replied:
i think i had seen an odd topic on a project which had a custom kafka topic viewer where the offsets were all even numbers (or maybe multiples of 4) for a sequence of messages, but that might have just been a ui bug; will leave it out for now

Scanteianu replied:
the main point is if you seek to offset 5, you'll also get offset 5 back (as opposed to the next message being 6)

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,53 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+  @Test
+  def testFetchOutOfRangeOffsetResetConfigEarliest(): Unit = {
+    this.consumerConfig.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
+    // ensure no records arrive before poll is called so that the offset actually gets reset
+    this.consumerConfig.setProperty(ConsumerConfig.FETCH_MAX_WAIT_MS_CONFIG, "0")
```
I think this comment is not clear. Maybe 
// ensure no in-flight fetch request so that the offset can be reset immediately

showuon commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -694,6 +694,53 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+  @Test
+  def testFetchOutOfRangeOffsetResetConfigLatest(): Unit = {
+    this.consumerConfig.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest")
+    // ensure no records arrive before poll is called so that the offset actually gets reset
+    this.consumerConfig.setProperty(ConsumerConfig.FETCH_MAX_WAIT_MS_CONFIG, "0")
```
ditto

showuon reviewed (commented):
@Scanteianu , thanks for the update. Left some more comments.

Scanteianu committed:
address further comments from @showuon

Scanteianu commented:
> > @Scanteianu , thanks for the update. Left some comments.
> 
> thanks for the comments @showuon , addressed!

thanks @showuon , addressed these as well

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

showuon reviewed (approved):
LGTM! Thanks for the contribution. 
Let's check the jenkins test results before merging it.

Scanteianu commented:
> LGTM! Thanks for the contribution. Let's check the jenkins test results before merging it.

Thank you so much!

The jenkins failing tests seem completely unrelated:

Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testReplicationIsCreatingTopicsUsingProvidedForwardingAdmin()

Build / JDK 17 and Scala 2.13 / org.apache.kafka.common.network.SslTransportLayerTest.[2] tlsProtocol=TLSv1.2, useInlinePem=true

org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testOffsetSyncsTopicsOnTarget()

I also started a discussion about the build stability on the mailing list and am trying to see what it might take to get a successful build by periodically updating this pr: https://github.com/apache/kafka/pull/12906/checks?check_run_id=10088101960 (which basically just disables failing tests, with the goal being to get a set of tests which, if fixed or removed, would achieve a regularly passing build)

Scanteianu committed:
Merge branch 'trunk' of https://github.com/apache/kafka into offsetDocumentation

Scanteianu commented:
@showuon jdk17/scala 2.13 passed, and the failures in the other runs seem unrelated

showuon commented:
Failed tests are unrelated and also fail in trunk build:
```
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.clients.consumer.StickyAssignorTest.testLargeAssignmentAndGroupWithNonEqualSubscription()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.clients.consumer/StickyAssignorTest/Build___JDK_8_and_Scala_2_12___testLargeAssignmentAndGroupWithNonEqualSubscription__/)
    [Build / JDK 8 and Scala 2.12 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testCreatePartitionsUseProvidedForwardingAdmin()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsWithCustomForwardingAdminIntegrationTest/Build___JDK_8_and_Scala_2_12___testCreatePartitionsUseProvidedForwardingAdmin__/)
    [Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.SmokeTestDriverIntegrationTest.[1] true](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.streams.integration/SmokeTestDriverIntegrationTest/Build___JDK_8_and_Scala_2_12____1__true/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testOffsetSyncsTopicsOnTarget()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.connect.mirror.integration/IdentityReplicationIntegrationTest/Build___JDK_11_and_Scala_2_13___testOffsetSyncsTopicsOnTarget__/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testNoCheckpointsIfNoRecordsAreMirrored()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.connect.mirror.integration/IdentityReplicationIntegrationTest/Build___JDK_11_and_Scala_2_13___testNoCheckpointsIfNoRecordsAreMirrored__/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testOneWayReplicationWithAutoOffsetSync()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.connect.mirror.integration/IdentityReplicationIntegrationTest/Build___JDK_11_and_Scala_2_13___testOneWayReplicationWithAutoOffsetSync__/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest.testReplication()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationBaseTest/Build___JDK_11_and_Scala_2_13___testReplication__/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest.testReplicationWithEmptyPartition()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationBaseTest/Build___JDK_11_and_Scala_2_13___testReplicationWithEmptyPartition__/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.BlockingConnectorTest.testBlockInSinkTaskStart](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.connect.integration/BlockingConnectorTest/Build___JDK_11_and_Scala_2_13___testBlockInSinkTaskStart/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.BlockingConnectorTest.testBlockInConnectorStop](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.connect.integration/BlockingConnectorTest/Build___JDK_11_and_Scala_2_13___testBlockInConnectorStop/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest.testBrokerCoordinator](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.connect.integration/ConnectWorkerIntegrationTest/Build___JDK_11_and_Scala_2_13___testBrokerCoordinator/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.[1] Type=Raft-CoReside, Name=testDescribeQuorumReplicationSuccessful, MetadataVersion=3.4-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_11_and_Scala_2_13____1__Type_Raft_CoReside__Name_testDescribeQuorumReplicationSuccessful__MetadataVersion_3_4_IV0__Security_PLAINTEXT/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.[2] Type=Raft-Distributed, Name=testDescribeQuorumReplicationSuccessful, MetadataVersion=3.4-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_11_and_Scala_2_13____2__Type_Raft_Distributed__Name_testDescribeQuorumReplicationSuccessful__MetadataVersion_3_4_IV0__Security_PLAINTEXT/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.[3] Type=Raft-CoReside, Name=testDescribeQuorumReplicationSuccessful, MetadataVersion=3.4-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_11_and_Scala_2_13____3__Type_Raft_CoReside__Name_testDescribeQuorumReplicationSuccessful__MetadataVersion_3_4_IV0__Security_PLAINTEXT/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.[4] Type=Raft-Distributed, Name=testDescribeQuorumReplicationSuccessful, MetadataVersion=3.4-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_11_and_Scala_2_13____4__Type_Raft_Distributed__Name_testDescribeQuorumReplicationSuccessful__MetadataVersion_3_4_IV0__Security_PLAINTEXT/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.[5] Type=Raft-CoReside, Name=testDescribeQuorumReplicationSuccessful, MetadataVersion=3.4-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_11_and_Scala_2_13____5__Type_Raft_CoReside__Name_testDescribeQuorumReplicationSuccessful__MetadataVersion_3_4_IV0__Security_PLAINTEXT/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.[6] Type=Raft-Distributed, Name=testDescribeQuorumReplicationSuccessful, MetadataVersion=3.4-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_11_and_Scala_2_13____6__Type_Raft_Distributed__Name_testDescribeQuorumReplicationSuccessful__MetadataVersion_3_4_IV0__Security_PLAINTEXT/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.[1] Type=Raft-CoReside, Name=testDescribeQuorumStatusSuccessful, MetadataVersion=3.4-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_11_and_Scala_2_13____1__Type_Raft_CoReside__Name_testDescribeQuorumStatusSuccessful__MetadataVersion_3_4_IV0__Security_PLAINTEXT/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.[2] Type=Raft-Distributed, Name=testDescribeQuorumStatusSuccessful, MetadataVersion=3.4-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_11_and_Scala_2_13____2__Type_Raft_Distributed__Name_testDescribeQuorumStatusSuccessful__MetadataVersion_3_4_IV0__Security_PLAINTEXT/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.[3] Type=Raft-CoReside, Name=testDescribeQuorumStatusSuccessful, MetadataVersion=3.4-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_11_and_Scala_2_13____3__Type_Raft_CoReside__Name_testDescribeQuorumStatusSuccessful__MetadataVersion_3_4_IV0__Security_PLAINTEXT/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.[5] Type=Raft-CoReside, Name=testDescribeQuorumStatusSuccessful, MetadataVersion=3.4-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_11_and_Scala_2_13____5__Type_Raft_CoReside__Name_testDescribeQuorumStatusSuccessful__MetadataVersion_3_4_IV0__Security_PLAINTEXT/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.[6] Type=Raft-Distributed, Name=testDescribeQuorumStatusSuccessful, MetadataVersion=3.4-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12753/33/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_11_and_Scala_2_13____6__Type_Raft_Distributed__Name_testDescribeQuorumStatusSuccessful__MetadataVersion_3_4_IV0__Security_PLAINTEXT/)
```
